www_src/pages/section-2/grid.jsx

```diff
@@ -0,0 +1,382 @@
+    }
+
+    this.animateCamera(true);
+  },
+  generateGrid: function (width, height) {
+    var grid = [];
+
+    for (var h = 0; h < height; h++) {
+      grid.push([]);
+
+      for (var w = 0; w < width; w++) {
+        grid[h].push(null);
+      }
+    }
+
+    grid[Math.floor(height / 2)][Math.floor(width / 2)] = 'EMPTY';
+
+    return grid;
+  },
+  getInitialState: function () {
+    // 20 x 20 seems to be max before display glitches
+    var layout = this.generateGrid(3, 3);
+
+    return {
+      layout: layout
+    }
+  },
+  addPageClick: function (event) {
+    var self = this;
+
+    var newLayout = this.state.layout;
+
+    newLayout[event.y][event.x] = 'EMPTY';
+
+    var width = newLayout[0].length;
+    var height = newLayout.length;
+
+    // The new page tile's x and y coords may shift if the grid is expanded
+    var newPageX = event.x;
+    var newPageY = event.y;
+
+    // Detect what grid boundaries the given tile touches
+    function edgeDetect(x, y) {
+      var boundaries = [];
+
+      if (x === 0) {
+        boundaries.push('left');
+      }
+
+      if (x === newLayout[0].length - 1) {
+        boundaries.push('right');
+      }
+
+      if (y === 0) {
+        boundaries.push('top');
+      }
+
+      if (y === newLayout.length - 1) {
+        boundaries.push('bottom');
+      }
+
+      return boundaries;
+    }
+
+    // Expand the grid in the specified direction
+    function expandGrid(direction) {
+      function buildEmptyRow(length) {
+        var row = [];
+
+        for (var i = 0; i < length; i++) {
+          row.push(null);
+        }
+
+        return row;
+      }
+
+      var expand = {
+        left: function () {
+          newLayout.forEach(function (row, index) {
+            row.unshift(null);
+          });
+
+          newPageX++;
+        },
+        right: function () {
+          newLayout.forEach(function (row, index) {
+            row.push(null);
+          });
+        },
+        top: function () {
+          newLayout.unshift(buildEmptyRow(newLayout[0].length));
+          newPageY++;
+        },
+        bottom: function () {
+          newLayout.push(buildEmptyRow(newLayout[0].length));
+        }
+      }
+
+      expand[direction]();
+    }
+
+    // Expand grid if necessary
+    if (edgeDetect(event.x, event.y).length) {
+      edgeDetect(event.x, event.y).forEach(function (direction, index, array) {
+        expandGrid(direction);
+      });
+    }
+
+    this.setState({
+      layout: newLayout
+    }, function () {
+      self.calculateCameraState(newPageX, newPageY, self.displayState.pagesWide, true);
+      self.animateCamera(false);
+    });
+  },
+  componentDidUpdate: function () {
+    var elGrid = this.getDOMNode();
+
+    // Force fixed width & height on grid
+    elGrid.style.width = this.gridWidth + 'px';
+    elGrid.style.height = this.gridHeight + 'px';
+  },
+  componentDidMount: function () {
+    this.gridWidth = undefined;
```

Comment:
setting all of these to undefined seems less useful than setting them to "defaults" instead (like zeroes across the board).